Statistics epic 2 3

Depechist opened the pull request:
Архитектура MVVM, верстка кодом

Задачи/Декомпозиция - https://vadamask.notion.site/a7d9bb129d8d4b9f9022d04de6a8bfd5?pvs=4
Доска - https://vadamask.notion.site/88b63082858c472ba96a17224134b9b9?v=0f49dd80115a48b2899b792c70b54117&pvs=4

Depechist committed:
Screen UI

- Сверстан экран Пользователя (данные захардкодены)

Depechist committed:
Network + WebView

- Данные о пользователе передаются на экран Пользователя через новый запрос UserById
- Моковые данные о Пользователе заменены на данные из сети
- По клику на кнопку сайта пользователя открывается WebView с сайтом указанным в загруженных данных о Пользователе

vadamask commented on FakeNFT/Scenes/Rating/RatingViewController.swift:
```diff
@@ -54,11 +54,12 @@ final class RatingViewController: UIViewController {
     }
 
     private func setupNavigationBar() {
+        navigationController?.navigationBar.tintColor = .yaBlack
         let image = UIImage(asset: Asset.sortButton)
         sortButton = UIBarButtonItem(image: image, style: .plain, target: self, action: #selector(sortButtonTapped))
```
Везде хорошо бы писать таким образом `let image = Asset.sortButton.image`

vadamask commented on FakeNFT/Scenes/User/CollectionCountCell.swift:
```diff
@@ -0,0 +1,73 @@
+//
+
+import UIKit
+import SnapKit
```
Лучше сортировать по алфавиту

vadamask commented on FakeNFT/Scenes/User/CollectionCountCell.swift:
```diff
@@ -0,0 +1,73 @@
+        let label = UILabel()
+        label.font = UIFont.bodyBold17
+        label.textColor = .textPrimary
+        label.text = NSLocalizedString(L10n.User.nftCollection, comment: "")
```
Здесь также можно просто `label.text = L10n.User.nftCollection`

vadamask commented on FakeNFT/Scenes/User/CollectionCountCell.swift:
```diff
@@ -0,0 +1,73 @@
+    }()
+
+    let customDisclosureIndicator: UIImageView = {
+        let imageView = UIImageView(image: UIImage(systemName: "chevron.right"))
```
Вложенную инициализацию лучше избегать

vadamask commented on FakeNFT/Scenes/User/UserViewModel.swift:
```diff
@@ -0,0 +1,44 @@
+        showLoading?()
+
+        let request = UserByIdRequest(id: userId)
+        networkClient.send(request: request, type: User.self) { [weak self] result in
+            self?.hideLoading?()
```
Чтобы избежать многих знаков вопроса, можно добавить `guard let self = self`

vadamask commented on FakeNFT/Scenes/User/UserViewController.swift:
```diff
@@ -0,0 +1,188 @@
+    init(userId: String) {
+        super.init(nibName: nil, bundle: nil)
+        self.viewModel = UserViewModel(networkClient: DefaultNetworkClient(), userId: userId)
+        viewModel?.userDetailsUpdated = { [weak self] user in
+            DispatchQueue.main.async {
+                self?.user = user
```
Здесь тоже можно добавить `guard let self = self`

vadamask commented on FakeNFT/Scenes/User/CollectionCountCell.swift:
```diff
@@ -0,0 +1,73 @@
+
+    private func setupConstraints() {
+        titleLabel.snp.makeConstraints { make in
+            make.top.equalToSuperview().offset(16)
+            make.leading.equalToSuperview().offset(16)
+            make.centerY.equalTo(contentView.snp.centerY)
```
Необязательно писать offset, можно написать `equalTo(16)` и snapkit поймет, что это ближайшая superview. 
Также и с центром - достаточно написать `equalTo(contentView)`

vadamask commented on FakeNFT/Scenes/User/UserViewController.swift:
```diff
@@ -0,0 +1,188 @@
+        websiteButton.layer.borderWidth = 1.0
+        websiteButton.titleLabel?.font = UIFont.caption15
+        websiteButton.setTitleColor(.textPrimary, for: .normal)
+        websiteButton.setTitle(NSLocalizedString(L10n.User.visitWebSite, comment: ""), for: .normal)
```
Нужно просто `L10n.User.visitWebSite`

vadamask reviewed (commented):
Отличная работа. Тебе стоит почитать официальную документацию про snapkit, там указаны все фишки :)

Depechist committed:
Refactoring

- Небольшой рефакторинг в классе UserViewController для улучшения читаемости и улучшения безопасности опционалов
- Небольшой рефакторинг после замечаний на ревью в команде

Nikitaaeee commented on FakeNFT/Scenes/User/CollectionCountCell.swift:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private func setupConstraints() {
+        titleLabel.snp.makeConstraints { make in
+            make.top.equalTo(16) // = make.top.equalToSuperview().offset(16)
+            make.leading.equalTo(16)
+            make.centerY.equalTo(contentView) // = make.centerY.equalTo(contentView.snp.centerY)
+        }
```
Если закомментированный код не будет использоваться - можно будет удалить

Nikitaaeee commented on FakeNFT/Scenes/User/UserViewController.swift:
```diff
@@ -0,0 +1,196 @@
+//  Created by Artem Adiev on 09.11.2023.
+//
+
+import UIKit
+import SnapKit
+import Kingfisher
+import ProgressHUD
```
Тут тоже можно расположить импорты в алфавитном порядке

Nikitaaeee commented on FakeNFT/Scenes/User/UserViewController.swift:
```diff
@@ -0,0 +1,196 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        viewModel?.showLoading = {
+            ProgressHUD.show()
+        }
+
+        viewModel?.hideLoading = {
+            ProgressHUD.dismiss()
+        }
```
Я заметил, что ты импортируешь **ProgressHUD** в каждом файле, где он используется. 
Хотел бы предложить создать отдельный файл - _менеджер_ для работы с индикаторами загрузки. 

Менеджер будет инкапсулировать всю логику, связанную с отображением и скрытием индикатора загрузки, и одновременно упростит код. 😊

Nikitaaeee commented on FakeNFT/Scenes/User/UserViewController.swift:
```diff
@@ -0,0 +1,196 @@
+        setupConstraints()
+    }
+
+    @objc func websiteButtonTapped() {
+        guard let url = userWebSiteUrl else {
+            print("Неверный URL")
+            return
+        }
+        let userWebViewController = UserWebViewController(url: url)
+        present(userWebViewController, animated: true)
+    }
```
Кажется, этому методу можно добавить **private**

Nikitaaeee commented on FakeNFT/Scenes/User/UserWebViewController.swift:
```diff
@@ -0,0 +1,39 @@
+
+import UIKit
+import WebKit
+
+class UserWebViewController: UIViewController {
+    // MARK: - Properties
```
Можно сделать класс **final**

Depechist committed:
Refactoring 2

- Небольшой рефакторинг после ревью
